src/main/kotlin/it/pagopa/wallet/services/WalletEventSinksService.kt

```diff
@@ -0,0 +1,67 @@
+package it.pagopa.wallet.services
+
+import it.pagopa.wallet.audit.LoggedAction
+import it.pagopa.wallet.config.properties.RetrySavePolicyConfig
+import it.pagopa.wallet.repositories.LoggingEventRepository
+import java.time.Duration
+import org.slf4j.LoggerFactory
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.boot.context.event.ApplicationReadyEvent
+import org.springframework.context.ApplicationListener
+import org.springframework.stereotype.Service
+import reactor.core.publisher.Flux
+import reactor.core.publisher.Mono
+import reactor.core.publisher.Sinks
+import reactor.core.scheduler.Schedulers
+import reactor.util.retry.Retry
+
+@Service
+class WalletEventSinksService(
+    @Autowired private val loggingEventRepository: LoggingEventRepository,
+    @Autowired private val retrySavePolicyConfig: RetrySavePolicyConfig,
+    private val walletEventSink: Sinks.Many<LoggedAction<*>> =
+        Sinks.many().unicast().onBackpressureBuffer()
+) : ApplicationListener<ApplicationReadyEvent> {
+    private val logger = LoggerFactory.getLogger(javaClass)
+
+    fun <T : Any> tryEmitEvent(loggedAction: LoggedAction<T>): Mono<LoggedAction<T>> =
+        Mono.fromCallable { walletEventSink.tryEmitNext(loggedAction) }
+            .flatMap {
+                if (it.isFailure)
+                    Mono.error(
+                        RuntimeException(
+                            "Sink failed to emit new wallet event: error code ${it.name}"
+                        )
+                    )
+                else Mono.just(loggedAction)
+            }
```

Comment:
```suggestion
            .map {
                it.orThrow()
                loggedAction
            }
```

this could be simplified as follow.
moreover this way we propagate the original exception in the chain instead of a RuntimeException

Reply:
We could use `orThrow` method but I preferred to chose the _reactor_ way since also the [documentation](https://projectreactor.io/docs/core/release/api/reactor/core/publisher/Sinks.EmitResult.html#orThrow--) of that method warns about throwing an exception in a reactor pipeline

Reply:
You are right but I think that the linked part of the documentation refers to the fact that in a reactive chain no part of the code should throw an Exception but map it to a Mono.error (that is one of the way to propagate an error in the reactive chain.
```code
So the following is an error because it explicitly throw an exception
fun fooWrong(a: String?): Mono<String> {
    return if (a == null) {
        throw RuntimeException("null input string")
    } else {
        Mono.just(a)
    }
}

where you should rather return a Mono.error instance

fun fooRight(a: String?): Mono<String> {
    return if (a == null) {
        Mono.error(RuntimeException("null input string"))
    } else {
       Mono.just(a)
    }
}
```

throwing an exception in a map method result in the Mono to be mapped to a Mono.error (propagating the error as an onError signal)

```code
fun main() {
    println(Mono
        .just("")
        .map {
            if (it.isBlank()) {
                throw RuntimeException("string is empty")
            }
            it
        }.doOnError { println("on error called, exception message: [${it.message}]") }
        .onErrorReturn("fallback")
        .block())
}

the above code snippet will print two lines:
on error called, exception message: [string is empty]
fallback
```

Reply:
Done in [0c11f9c](https://github.com/pagopa/pagopa-wallet-service/pull/195/commits/0c11f9c57c45a1f965b62419b12c3e5912d280e1)